inspirehep/dojson/hep/fields/bd0xx.py

```diff
@@ -0,0 +1,331 @@
+
+import re
+
+import pycountry
+
+from dojson import utils
+from idutils import is_doi, is_handle, normalize_doi
+
+from inspire_schemas.utils import load_schema
+from inspirehep.utils.helpers import force_force_list
+
+from ..model import hep, hep2marc
+from ...utils import force_single_element
+
+
+RE_LANGUAGE = re.compile('\/| or | and |,|=|\s+')
+
+
+@hep.over('isbns', '^020..')
+@utils.for_each_value
+def isbns(self, key, value):
+    def _get_medium(value):
+        def _normalize(medium):
+            schema = load_schema('hep')
+            valid_mediums = schema['properties']['isbns']['items']['properties']['medium']['enum']
+
+            medium = medium.lower().replace('-', '').replace(' ', '')
+            if medium in valid_mediums:
+                return medium
+            elif medium == 'ebook':
+                return 'online'
+            elif medium == 'paperback':
+                return 'softcover'
+
+            return ''
+
+        medium = force_single_element(value.get('b', ''))
+        normalized_medium = _normalize(medium)
+
+        return normalized_medium
+
+    def _get_isbn(value):
+        def _normalize(isbn):
+            return isbn.lower().replace('-', '')
+
+        isbn = force_single_element(value.get('a', ''))
+        normalized_isbn = _normalize(isbn)
+
+        return normalized_isbn
+
+    return {
+        'medium': _get_medium(value),
+        'value': _get_isbn(value),
+    }
+
+
+@hep2marc.over('020', 'isbns')
+@utils.for_each_value
+def isbns2marc(self, key, value):
+    return {
+        'a': value.get('value'),
+        'b': value.get('medium'),
+    }
+
+
+@hep.over('dois', '^0247.')
+def dois(self, key, value):
+    """Populate the ``dois`` key.
+
+    Also populates the ``persistent_identifiers`` key through side effects.
+    """
+    def _get_first_non_curator_source(sources):
+        sources_without_curator = [el for el in sources if el.upper != 'CURATOR']
+        return force_single_element(sources_without_curator)
+
+    def _is_doi(id_, type_):
+        return (not type_ or type_.upper() == 'DOI') and is_doi(id_)
+
+    def _is_handle(id_, type_):
+        return (not type_ or type_.upper() in ('DOI', 'HDL')) and is_handle(id_)
+
+    dois = self.get('dois', [])
+    persistent_identifiers = self.get('persistent_identifiers', [])
+
+    values = force_force_list(value)
+    for value in values:
+        id_ = force_single_element(value.get('a', ''))
+        material = force_single_element(value.get('q', ''))
+        schema = force_single_element(value.get('2', ''))
+
+        sources = force_force_list(value.get('9'))
+        source = _get_first_non_curator_source(sources)
+
+        if _is_doi(id_, schema):
+            dois.append({
+                'material': material,
+                'source': source,
+                'value': normalize_doi(id_),
+            })
+        else:
+            schema = 'HDL' if _is_handle(id_, schema) else schema
```

Comment:
* If I understand this correctly, `0247_2:DOI` with invalid DOI gets put into `persistent_identifiers`, which fails because `schema` is an [enum](https://github.com/inspirehep/inspire-schemas/blob/4d88e8d128a3ea0d7465581fd72f3e4965f6b50f/inspire_schemas/records/hep.json#L598). Is that what should happen?
* Don't you similarly want to check for `URN`?